File: 백준/Silver/9012. 괄호/괄호.py
import sys
from collections import deque

input = sys.stdin.readline

N= int(input())

for i in range(N):
    X = deque(input().rstrip())
    stack = []
    check = False

    while X:
        if X[0] == '(':
            stack.append(1)
            X.popleft()
        elif X[0] ==')':
            if len(stack)==0:
                break
            else :
                stack.pop()
                X.popleft()

    if len(X) == 0 and len(stack) == 0:
        check = True

    if check :
        print("YES")
    else :
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        for(int i=0;i<N;i++){
            Stack<Character> stack = new Stack<>();

            String input = br.readLine();
            Boolean isVPS = true;
            for(int j=0;j<input.length();j++){
                Character temp = input.charAt(j);

                if(temp == '(') stack.push('(');
                else{
                    if(stack.isEmpty()){
                        isVPS = false;
                        break;
                    }else stack.pop();
                }

            }

            if(isVPS && stack.isEmpty()) System.out.println("YES");
            else System.out.println("NO");
        }


    }
}